libsolidity/ast/Types.cpp

```diff
@@ -451,44 +660,93 @@ TypePointer IntegerConstantType::binaryOperatorResult(Token::Value _operator, Ty
 			return TypePointer();
 		return intType->binaryOperatorResult(_operator, _other);
 	}
+	else if (_other->category() == Category::Fixed)
+	{
+		shared_ptr<FixedType const> fixType = fixedType();
+		if (!fixType)
+			return TypePointer();
+		return fixType->binaryOperatorResult(_operator, _other);
+	}
 	else if (_other->category() != category())
 		return TypePointer();
 
-	IntegerConstantType const& other = dynamic_cast<IntegerConstantType const&>(*_other);
+	ConstantNumberType const& other = dynamic_cast<ConstantNumberType const&>(*_other);
 	if (Token::isCompareOp(_operator))
 	{
-		shared_ptr<IntegerType const> thisIntegerType = integerType();
-		shared_ptr<IntegerType const> otherIntegerType = other.integerType();
-		if (!thisIntegerType || !otherIntegerType)
-			return TypePointer();
-		return thisIntegerType->binaryOperatorResult(_operator, otherIntegerType);
+		if (isInteger())
+		{
+			shared_ptr<IntegerType const> thisIntegerType = integerType();
+			shared_ptr<IntegerType const> otherIntegerType = other.integerType();
+			if (!thisIntegerType || !otherIntegerType)
+				return TypePointer();
+			return thisIntegerType->binaryOperatorResult(_operator, otherIntegerType);
+		}
+		else
+		{
+			shared_ptr<FixedType const> thisFixedType = fixedType();
+			shared_ptr<FixedType const> otherFixedType = other.fixedType();
+			if (!thisFixedType || !otherFixedType)
+				return TypePointer();
+			return thisFixedType->binaryOperatorResult(_operator, otherFixedType);
+		}		
 	}
 	else
 	{
 		bigint value;
+		short scale = m_scalingFactor;
+		Modifier modifier = isInteger() && other.isInteger() ? Modifier::Integer : Modifier::Fixed;
 		switch (_operator)
 		{
+		//bit operations will only be enabled for integers and fixed types that resemble integers
 		case Token::BitOr:
+			if (modifier == Modifier::Fixed)
+				if(rightOfRadix() != bigint(0) || other.rightOfRadix() != bigint(0))
+					BOOST_THROW_EXCEPTION(InternalCompilerError() << errinfo_comment("Cannot perform bit operations on non integer fixed type."));
 			value = m_value | other.m_value;
 			break;
 		case Token::BitXor:
+			if (modifier == Modifier::Fixed)
+				if(rightOfRadix() != bigint(0) || other.rightOfRadix() != bigint(0))
+					BOOST_THROW_EXCEPTION(InternalCompilerError() << errinfo_comment("Cannot perform bit operations on non integer fixed type."));
 			value = m_value ^ other.m_value;
 			break;
 		case Token::BitAnd:
+			if (modifier == Modifier::Fixed)
+				if(rightOfRadix() != bigint(0) || other.rightOfRadix() != bigint(0))
+					BOOST_THROW_EXCEPTION(InternalCompilerError() << errinfo_comment("Cannot perform bit operations on non integer fixed type."));
 			value = m_value & other.m_value;
 			break;
 		case Token::Add:
-			value = m_value + other.m_value;
+			if (modifier == Modifier::Fixed)
+			{
+				bigint leftSide = leftOfRadix() + other.leftOfRadix();
```

Comment:
This looks like we actually do not need a special case for fixed point arithmetics.